From: Vincent Whitchurch <vincent.whitchurch@axis.com>

We currently do not respect off_on_delay the first time we turn on a
regulator.  This is problematic since the regulator could have been
turned off by the bootloader, or it could it have been turned off during
the probe of the regulator driver (such as when regulator-fixed requests
the enable GPIO), either of which could potentially have happened less
than off_on_delay microseconds ago before the first time a client
requests for the regulator to be turned on.

We can't know exactly when the regulator was turned off, but initialise
off_on_delay to the current time when registering the regulator, so that
we guarantee that we respect the off_on_delay in all cases.

Signed-off-by: Vincent Whitchurch <vincent.whitchurch@axis.com>
Link: https://lore.kernel.org/r/20210422083044.11479-1-vincent.whitchurch@axis.com
Signed-off-by: Mark Brown <broonie@kernel.org>
---
 drivers/regulator/core.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/regulator/core.c b/drivers/regulator/core.c
index 7fd793d8536c..bfd93524cdca 100644
--- a/drivers/regulator/core.c
+++ b/drivers/regulator/core.c
@@ -1404,6 +1404,8 @@ static int set_machine_constraints(struct regulator_dev *rdev)

                if (rdev->constraints->always_on)
                        rdev->use_count++;
+       } else if (rdev->desc->off_on_delay) {
+               rdev->last_off_jiffy = jiffies;
        }

        print_constraints(rdev);
--
2.33.0
